cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java

```diff
@@ -146,79 +146,86 @@ public KafkaLogWriterPlugin(CConfiguration cConfig, FileMetaDataManager fileMeta
-  public void doProcess(KafkaLogEvent event) {
+  public void doProcess(Iterator<KafkaLogEvent> events) {
+    // Get the timestamp from the first event to compute time bucket
+    PeekingIterator<KafkaLogEvent> peekingIterator = Iterators.peekingIterator(events);
+    KafkaLogEvent firstEvent = peekingIterator.peek();
 
-    LoggingContext loggingContext = event.getLoggingContext();
-    ILoggingEvent logEvent = event.getLogEvent();
     try {
-      // Compute the bucket number for the current event
-      long key = logEvent.getTimeStamp() / eventBucketIntervalMs;
+      // Compute the bucket number for the first event
+      long firstKey = firstEvent.getLogEvent().getTimeStamp() / eventBucketIntervalMs;
 
       // Sleep while we can add the entry
       while (true) {
         // Get the oldest bucket in the table
         long oldestBucketKey;
         synchronized (messageTable) {
           SortedSet<Long> rowKeySet = messageTable.rowKeySet();
-          if (rowKeySet.isEmpty()) {
-            // Table is empty so go ahead and add the current event in the table
+          oldestBucketKey = rowKeySet.isEmpty() ? System.currentTimeMillis() : rowKeySet.first();
+
+          // If the current event falls in the bucket number which is in window [oldestBucketKey, oldestBucketKey+8]
+          // then we can add the event to message table
+          if (firstKey <= (oldestBucketKey + maxNumberOfBucketsInTable)) {
+            while (peekingIterator.hasNext()) {
+              KafkaLogEvent event = peekingIterator.next();
+              LoggingContext loggingContext = event.getLoggingContext();
+              long key = event.getLogEvent().getTimeStamp() / eventBucketIntervalMs;
+
+              Map.Entry<Long, List<KafkaLogEvent>> entry =
+                messageTable.get(key, loggingContext.getLogPathFragment(logBaseDir));
+              List<KafkaLogEvent> msgList;
+              if (entry == null) {
+                long eventArrivalBucketKey = System.currentTimeMillis() / eventBucketIntervalMs;
+                msgList = Lists.newArrayList();
+                messageTable.put(key, loggingContext.getLogPathFragment(logBaseDir),
+                                 new AbstractMap.SimpleEntry<>(eventArrivalBucketKey, msgList));
+              } else {
+                msgList = messageTable.get(key, loggingContext.getLogPathFragment(logBaseDir)).getValue();
+              }
+              msgList.add(new KafkaLogEvent(event.getGenericRecord(), event.getLogEvent(), loggingContext,
+                                            event.getPartition(), event.getNextOffset()));
+            }
             break;
           }
-          oldestBucketKey = rowKeySet.first();
         }
 
-        // If the current event falls in the bucket number which is not in window [oldestBucketKey, oldestBucketKey+8]
+        // Cannot insert event into message table
+        // since there are still maxNumberOfBucketsInTable buckets that need to be processed
```

Comment:
One thread is writing to the table while another thread is reading and removing events from the table. To prevent OOM, we have a limit of 8 buckets in memory.